src/Evolution/Particles/MonteCarlo/Scattering.cpp

```diff
@@ -3,11 +3,144 @@
+          TargetF / (1. - exp(-cTau));
+    }
+  }
+}
+
+void DiffusionPrecomputeForElement(
+    gsl::not_null<DataVector*> prefactor_diffusion_time_vector,
+    gsl::not_null<DataVector*> prefactor_diffusion_four_velocity,
+    gsl::not_null<DataVector*> prefactor_diffusion_time_step,
+    const Scalar<DataVector>& lorentz_factor,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& lower_spatial_four_velocity,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& shift,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric) {
+  // g_tt, stored in prefactor_diffusion_time_step
+  // g_tt = \alpha - \beta^i \beta_i
+  *prefactor_diffusion_time_step =
+      -(get(lapse) * get(lapse)) +
+      get(dot_product(shift, shift, spatial_metric));
+  // u_t stored in prefactor_diffusion_four_velocity
+  // u_t = - W * (lapse - beta^i v_i)
+  *prefactor_diffusion_four_velocity =
+      get(lorentz_factor) * get(lapse) * (-1.0) +
+      get(dot_product(shift, lower_spatial_four_velocity));
+  // Precompute u_t^2 + g_tt
+  *prefactor_diffusion_time_vector = (*prefactor_diffusion_four_velocity) *
+                                         (*prefactor_diffusion_four_velocity) +
+                                     (*prefactor_diffusion_time_step);
+
+  for (size_t i = 0; i < prefactor_diffusion_time_vector->size(); i++) {
+    // First, deal with case with small transport velocity. In that case,
+    // Eq 32 is singular, but we also do not need to move packets with
+    // the fluid in the diffusion step (as the fluid is not moving on
+    // the grid). Set prefactor_diffusion_time_step to -1, which we will
+    // check for when performing diffusion.
+    if (fabs((*prefactor_diffusion_time_vector)[i]) < 1.e-10) {
+      (*prefactor_diffusion_time_step)[i] = -1.0;
+      (*prefactor_diffusion_time_vector)[i] = 1.0;
+      (*prefactor_diffusion_four_velocity)[i] = 1.0;
+      continue;
+    }
+    // Eq 31 stored in prefactor_diffusion_time_step
+    // A/B = - (u_t + sqrt(u_t^2 + g_tt))/g_tt
+    (*prefactor_diffusion_time_step)[i] =
+        -((*prefactor_diffusion_four_velocity)[i] +
+          sqrt((*prefactor_diffusion_time_vector)[i])) /
+        (*prefactor_diffusion_time_step)[i];
+    // Eq. 32 for neutrinos of 1MeV energy (the result scale linearly with
+    // energy, and can thus be modified as needed during packet evolution) B = 1
+    // / (1 - (A/B) u_t)
+    (*prefactor_diffusion_four_velocity)[i] =
+        1.0 / (1.0 - (*prefactor_diffusion_time_step)[i] *
+                         (*prefactor_diffusion_four_velocity)[i]);
+    // A from Eq. 32 and 31
+    // A= (A/B) * B
+    (*prefactor_diffusion_time_vector)[i] =
+        (*prefactor_diffusion_time_step)[i] *
+        (*prefactor_diffusion_four_velocity)[i];
+    // Eq. 33
+    // prefactor_diffusion_time_step = (A + B u^t)/(B u^t) = 1 + A*alpha/(B*W)
+    (*prefactor_diffusion_time_step)[i] =
```

Comment:
since you have the temporary double A_over_B, you can just use it in this equation